module/zfs/zio.c

```diff
@@ -1212,6 +1212,12 @@ zfs_blkptr_verify(spa_t *spa, const blkptr_t *bp,
 			    bp, (longlong_t)BPE_GET_PSIZE(bp));
 		}
 		return (errors == 0);
+	} else if (BP_IS_HOLE(bp)) {
+		/* Holes are allowed (expected, even) to have no DVAs */
+	} else if (unlikely(!DVA_IS_VALID(&bp->blk_dva[0]))) {
+		/* Non-hole, non-embedded BPs _must_ have at least one DVA */
+		errors += zfs_blkptr_verify_log(spa, bp, blk_verify,
+		    "blkptr at %px has no valid DVAs", bp);
```

Comment:
I have no objections, but wonder if we could/should go deeper, checking other DVAs too, to make sure they are either empty or valid, or being forgiving there is OK.

Reply:
My feeling is that we should be forgiving, at least for reads, because we don't know what might be written to existing file systems.  Maybe parts of the DVAs are uninitialized, but current ZFS doesn't care?  We don't want to render those file systems unreadable.